esp/esp/program/controllers/autoscheduler/data_model.py

```diff
@@ -0,0 +1,318 @@
+                 "consider locking the offending class(es)."
+                 .format(violation)))
+
+
+class AS_ClassSection(object):
+    def __init__(self, teachers, duration, capacity,
+                 category, assigned_roomslots,
+                 section_id=0, parent_class_id=0,
+                 grade_min=7, grade_max=12,
+                 resource_requests=None):
+        self.id = section_id
+        self.parent_class = parent_class_id
+        # Duration, in hours.
+        self.duration = duration
+        # A list of teachers.
+        self.teachers = teachers
+        # Capacity, an int.
+        self.capacity = capacity
+        # Min and max grade, integers
+        self.grade_min = grade_min
+        self.grade_max = grade_max
+        # Category ID
+        self.category = category
+        # A sorted list of assigned roomslots.
+        self.assigned_roomslots = assigned_roomslots
+        for roomslot in assigned_roomslots:
+            roomslot.assigned_section = self
+            for teacher in self.teachers:
+                teacher.add_availability(roomslot.timeslot)
+        # Dict from restype names to AS_Restypes requested
+        self.resource_requests = resource_requests \
+            if resource_requests is not None else {}
+
+        # A hash of the initial state.
+        self.recompute_hash()
+        self.register_teachers()
+
+    def register_teachers(self):
+        """Makes sure that all teachers have this section listed in their taught
+        sections."""
+        for teacher in self.teachers:
+            teacher.taught_sections[self.id] = self
+
+    def assign_roomslots(self, roomslots, clear_existing=False):
+        if not clear_existing:
+            assert self.assigned_roomslots == [], \
+                    "Already assigned to roomslots"
+        else:
+            self.clear_roomslots()
+        self.assigned_roomslots = sorted(roomslots, key=lambda r: r.timeslot)
+        for roomslot in self.assigned_roomslots:
+            assert roomslot.assigned_section is None, \
+                    "Roomslot is occupied"
+            roomslot.assigned_section = self
+
+    def clear_roomslots(self):
+        for roomslot in self.assigned_roomslots:
+            roomslot.assigned_section = None
+        self.assigned_roomslots = []
+
+    def is_scheduled(self):
+        return len(self.assigned_roomslots) > 0
+
+    def recompute_hash(self):
+        self.initial_state = self.scheduling_hash()
+
+    def scheduling_hash(self):
+        """Creates a unique hash based on the timeslots and rooms assigned to
+        this section."""
+        meeting_times = sorted([(str(e.timeslot.start), str(e.timeslot.end))
+                                for e in self.assigned_roomslots])
+        rooms = sorted(list(set([r.room.name for r in
+                                self.assigned_roomslots])))
+        state_str = json.dumps([meeting_times, rooms])
+        return hashlib.md5(state_str).hexdigest()
```

Comment:
is hashing this complicatedly necessary? (it's probably fine but all other things being equal there are better hashes than md5, aren't there?)